graylog2-server/src/main/java/org/graylog2/migrations/V20180214093600_AdjustDashboardPositionToNewResolution.java

```diff
@@ -0,0 +1,123 @@
+/**
+ * This file is part of Graylog.
+ *
+ * Graylog is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Graylog is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.graylog2.migrations;
+
+import com.google.common.collect.Maps;
+
+import com.mongodb.BasicDBObject;
+import org.graylog.autovalue.WithBeanGetter;
+import org.graylog2.dashboards.Dashboard;
+import org.graylog2.dashboards.DashboardImpl;
+import org.graylog2.dashboards.DashboardService;
+import org.graylog2.plugin.cluster.ClusterConfigService;
+import org.graylog2.plugin.database.ValidationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.google.auto.value.AutoValue;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.time.ZonedDateTime;
+import java.util.Map;
+
+import javax.inject.Inject;
+
+import static java.lang.Integer.parseInt;
+
+/**
+ * Migration adjusting the position of dashboard widgets to the higher resolution of the
+ * grid layout.
+ */
+public class V20180214093600_AdjustDashboardPositionToNewResolution extends Migration {
+
+    private static final Logger LOG = LoggerFactory.getLogger(V20180214093600_AdjustDashboardPositionToNewResolution.class);
+    private final DashboardService dashboardService;
+    private final ClusterConfigService clusterConfigService;
+
+    @Inject
+    public V20180214093600_AdjustDashboardPositionToNewResolution(DashboardService dashboardService,
+                                                                ClusterConfigService clusterConfigService) {
+        this.dashboardService = dashboardService;
+        this.clusterConfigService = clusterConfigService;
+    }
+
+
+    public ZonedDateTime createdAt() { return ZonedDateTime.parse("2018-02-14T09:36:00Z"); }
+
+    @Override
+    public void upgrade() {
+        if (clusterConfigService.get(MigrationCompleted.class) != null) {
+            LOG.debug("Migration already completed!");
+            return;
+        }
+
+        Map<String, String> dashboardIds = Maps.newHashMap();
+        for (Dashboard dashboard : dashboardService.all()) {
+            Map<String, Object> fields = dashboard.getFields();
+            BasicDBObject positions = (BasicDBObject) fields.get(DashboardImpl.EMBEDDED_POSITIONS);
```

Comment:
Having to use parts of an implementation class (`Dashboard_Impl_`) on an upper abstraction level is always a sign that there is missing functionality in the implemented interface. I think we should implement the functionality to get the position of a widget in a structured/typed format (this way we can also get rid of all the following casts and direct field accesses), modify the positions and save them back using `DashboardService#updateWidgetPositions`.